Delete existing user_keys, if password is changed

schultz-it-solutions opened the pull request:
Pull Request for Issue #17825 .

### Summary of Changes
new method 'onUserBeforeSave'
deletes all #__user_keys database records for current user, if he changed his password (and this option is activated in the plugin configuration.

### Testing Instructions
1.   Login on one device with activated "remember-me" checkbox and close the browser again.
2.   Login on another device and change your password here.
3.   Go back to your first device and open the website again.

### Expected result
Website should not allow you to login without providing new password


### Actual result



### Documentation Changes Required

### additional changes in other files:
https://github.com/joomla/joomla-cms/compare/staging...schultz-it-solutions:patch-1 (Plugin XML definition)
https://github.com/joomla/joomla-cms/compare/staging...schultz-it-solutions:patch-3 (Language File)
https://github.com/joomla/joomla-cms/compare/staging...schultz-it-solutions:patch-4 (Language File)

schultz-it-solutions committed:
Delete existing user_keys, if password is changed

dgrammatiko commented on plugins/system/remember/remember.php:
```diff
@@ -94,4 +94,61 @@ public function onUserLogout($user, $options)
+	 *
+	 * @return    boolean
+	 *
+	 * @since   3.1
```
3.1 has been released few years ago :)
Please change this to `__DEPLOY_VERSION__`

dgrammatiko commented:
The code has some code styling issues:
```
--------------------------------------------------------------------------------
26s
57
FOUND 10 ERROR(S) AND 1 WARNING(S) AFFECTING 8 LINE(S)
26s
58
--------------------------------------------------------------------------------
26s
59
 100 | WARNING | Line exceeds 150 characters; contains 161 characters
26s
60
 115 | ERROR   | Please start your comment with a capital letter; found "//
26s
61
     |         | irelevant on new users"
26s
62
 116 | ERROR   | Expected "if (...)\n"; found "if (...) "
26s
63
 116 | ERROR   | Closing brace must be on a line by itself
26s
64
 117 | ERROR   | Please start your comment with a capital letter; found "//
26s
65
     |         | irelevant, because password was not changed by user"
26s
66
 118 | ERROR   | Expected "if (...)\n"; found "if (...) "
26s
67
 118 | ERROR   | Closing brace must be on a line by itself
26s
68
 119 | ERROR   | Please start your comment with a capital letter; found "//
26s
69
     |         | irelevant, because "resetting on pw change" is not activated"
26s
70
 120 | ERROR   | Expected "if (...)\n"; found "if (...) "
26s
71
 120 | ERROR   | Closing brace must be on a line by itself
26s
72
 149 | ERROR   | Please put a space between the // and the start of comment
26s
73
     |         | text; found "//$this->app->input->cookie->set($cookieName, '',
26s
74
     |         | 1, $this->app->get('cookie_path', '/'),
26s
75
     |         | $this->app->get('cookie_domain', ''));"
26s
76
--------------------------------------------------------------------------------
26s
77
UPGRADE TO PHP_CODESNIFFER 2.0 TO FIX ERRORS AUTOMATICALLY
26s
78
--------------------------------------------------------------------------------
```

brianteeman commented:
please add your other changes to this pr. you can do that by going to this branch in your fork and editing the files there. they will then be added automatically to this pull request

schultz-it-solutions committed:
corrected styling issues

wojsmol commented on plugins/system/remember/remember.php:
```diff
@@ -107,17 +108,26 @@ public function onUserLogout($user, $options)
 	 * @return    boolean
 	 *
-	 * @since   3.1
+	 * @since   3.7
```
`@since   __DEPLOY_VERSION__` as mentioned by @dgt41

schultz-it-solutions committed:
deploy version - as I said, this is my first pr

dgrammatiko commented on plugins/system/remember/remember.php:
```diff
@@ -94,4 +94,69 @@ public function onUserLogout($user, $options)
+		{
+			$this->app = JFactory::getApplication();
+		}
+		/*
```
new line after `}`

dgrammatiko commented on plugins/system/remember/remember.php:
```diff
@@ -94,4 +94,69 @@ public function onUserLogout($user, $options)
+		{
+			return true;
+		}
+		// Irrelevant, because password was not changed by user
```
new line after `}`

dgrammatiko commented on plugins/system/remember/remember.php:
```diff
@@ -94,4 +94,69 @@ public function onUserLogout($user, $options)
+		{
+			return true;
+		}
+		// Irrelevant, because "resetting on pw change" is not activated
```
new line after `}`

schultz-it-solutions committed:
pushing to patch-2

dgrammatiko commented on plugins/system/remember/remember.php:
```diff
@@ -94,4 +94,69 @@ public function onUserLogout($user, $options)
+				'security'
+			);
+		}
+		$this->app->enqueueMessage(JText::_('COM_USERS_PROFILE_SAVE_REMEMBERME_USERINFO'), 'notice'); 
```
new line after `}`

schultz-it-solutions committed:
newline after }

schultz-it-solutions committed:
push to patch-2

schultz-it-solutions committed:
push to patch-2

schultz-it-solutions committed:
Update en-GB.com_users.ini

schultz-it-solutions committed:
Update remember.php

Quy commented on administrator/language/en-GB/en-GB.plg_system_remember.ini:
```diff
@@ -5,3 +5,5 @@
 
 PLG_REMEMBER_XML_DESCRIPTION="Provides remember me functionality. The cookie authentication plugin must be enabled for this plugin to function."
 PLG_SYSTEM_REMEMBER="System - Remember Me"
+PLG_REMEMBER_RESET_DESC="Should the 'remember-me' functionality be reset (recommended), when a user changes his password? This will render any existing login cookies on other devices useless."
```
Sort keys in alpha order. 
Remove comma before `when`.
Remove `his` to be gender neutral.

mbabker replied:
Use this string please:

"When enabled, a user's login cookies will be invalidated when their password is reset."

schultz-it-solutions replied:
done, as per mbabker suggestion

schultz-it-solutions committed:
Update remember.xml

Quy commented on plugins/system/remember/remember.php:
```diff
@@ -94,4 +94,73 @@ public function onUserLogout($user, $options)
+		{
+			$db->setQuery($query)->execute();
+		}
+		
```
Remove blank line.

schultz-it-solutions replied:
done

Quy commented on administrator/language/en-GB/en-GB.plg_system_remember.ini:
```diff
@@ -5,3 +5,5 @@
 PLG_REMEMBER_XML_DESCRIPTION="Provides remember me functionality. The cookie authentication plugin must be enabled for this plugin to function."
 PLG_SYSTEM_REMEMBER="System - Remember Me"
+PLG_REMEMBER_RESET_DESC="Should the 'remember-me' functionality be reset (recommended), when a user changes his password? This will render any existing login cookies on other devices useless."
+PLG_REMEMBER_RESET_LABEL="reset on password change"
```
Capitalize each word.

schultz-it-solutions replied:
not sure about this, I suggest:

Reset on password change

Quy commented on plugins/system/remember/remember.php:
```diff
@@ -94,4 +94,73 @@ public function onUserLogout($user, $options)
+		}
+		
+		$this->app->enqueueMessage(JText::_('COM_USERS_PROFILE_SAVE_REMEMBERME_USERINFO'), 'notice'); 
+		return true;
```
Add a blank line before line 163.

schultz-it-solutions replied:
done

schultz-it-solutions committed:
configuration option in XML file

radio button option to activate/deactivate the "reset RememberMe" functionality on password-change.

Quy commented:
Please follow XML coding style:
https://developer.joomla.org/coding-standards/xml.html

mbabker commented on language/en-GB/en-GB.com_users.ini:
```diff
@@ -87,6 +87,7 @@ COM_USERS_PROFILE_PASSWORD2_DESC="Confirm your password."
 COM_USERS_PROFILE_PASSWORD2_LABEL="Confirm Password"
 COM_USERS_PROFILE_REGISTERED_DATE_LABEL="Registered Date"
 COM_USERS_PROFILE_SAVE_FAILED="Profile could not be saved: %s"
+COM_USERS_PROFILE_SAVE_REMEMBERME_USERINFO="Your password change was successful. For your security, we also deactivated any existing 'remember-me' logins on other devices."
```
I don't think we actually need to explicitly say "your remember me cookies are invalidated", I can't say I see this come up on other platforms where I have a remember me login in place, the device just knows the next time I use it to access a service I need to reauthenticate (well, the server tells it when it tries to automatically log in, but that's all transparent to the user).

schultz-it-solutions replied:
This might seem 'over-the-top', but it was intended to 'reassure the user, everything is settled'. We could ignore this message OR (which I would prefer) let the website administrator decide whether to show this message or not (another configuration option in the plugin).

dgrammatiko replied:
I'm also in favour of less clutter...

Bakual replied:
I would also not show the message.

schultz-it-solutions committed:
Update en-GB.plg_system_remember.ini

schultz-it-solutions committed:
hm...

schultz-it-solutions committed:
Update remember.php

Bakual commented:
Do we really need an option for that? I would just go without it.
If you really want that parameter, the default values in code and in XML will have to match to prevent unexpected behavior.

Bakual commented:
Another thing to consider: You put the code into the "System - Remember" plugin, but that is actually just a proxy plugin so the real "Authentication - Cookie" plugin can do its work. So this code should go into that authentication plugin instead.

mbabker commented:
If we aren't importing authentication plugins in the MVC for CRUD actions then this is the right plugin.

Bakual commented:
How about we import the authentication plugins instead? I'd say that should be easy and could be useful for other authentication plugins as well. I don't think it's a good idea to spread the logic of the cookie plugin (again) over multiple ones. If we can, it should be kept within the cookie one.

mbabker commented:
The authentication plugins shouldn't be handling CRUD actions (since some people seem to be overly concerned about what tasks specific plugin groups actually listen to, apparently we should have a "User - Remember Me" plugin for this task if you buy into their thinking 😉 ).

The authentication plugins in general aren't treated the same as every other plugin group, there are actually only two places in core where they are actually loaded as "real" plugins otherwise the Authentication class pulls the plugin objects out of the dispatcher and direct calls the methods (so not using the event system).  I would be hesitant to add more plugin based functionality to this plugin group since it is not treated as a correct plugin group (if we want to use it as a real plugin group the authentication system needs to be refactored to use proper events and not the crap logic it has now).

As far as where the code is, things are fine right now.

schultz-it-solutions commented:
On the "user message": I may be "over sensible" in this respect, because of the underlying story why this pr exists...
If you think this user message is "clutter", I am not oposing this!

Bakual commented on plugins/system/remember/remember.php:
```diff
@@ -94,4 +94,73 @@ public function onUserLogout($user, $options)
+	 * Method is called before user data is stored in the database
+	 * If activated in the configuration of the rememeber-me plugin, this method resets all #__user_keys for the current user
+	 * when the user changes his/her password, leaving any existing remember-me cookies on any devices useless!
+	 * This functionality was sadly inspired by the horrific events around Alice Ruggles death (see https://www.alicerugglestrust.org )
```
Please remove this reference to Alice Ruggles and especially also the URL as it serves no purpose within code.

schultz-it-solutions replied:
I removed the URL, but I would rather like to keep the reference, as this is the underlying reason why this functionality is developed in the first place.

Bakual replied:
Honestly, I never heard of her and even after reading that she got killed after being stalked I don't get the reference. So it seems quite useless to me.
Also, on the other hand it's quite obvious why we want to invalidate existing remember-me credentials. Personally I would just add a one-line comment like `Invalidate all existing remember-me cookies after a password change`.

schultz-it-solutions replied:
Nobody expects that any given person "has heard of" each and every stalking victim, this is not the point. However, one major element in this case was that the murderer actually hacked into her online accounts to follow her movements, using exactly this "remember-me" functionality (not on Joomla though).

Bakual commented on plugins/system/remember/remember.php:
```diff
@@ -94,4 +94,73 @@ public function onUserLogout($user, $options)
+		}
+
+		// Get the application if not done by JPlugin. This may happen during upgrades from Joomla 2.5.
+		if (!$this->app)
```
This isn't needed in that event to my knowledge as that event will not be fired **during** an upgrade.

schultz-it-solutions replied:
done

schultz-it-solutions committed:
Update remember.php

schultz-it-solutions committed:
XML styles

schultz-it-solutions committed:
commenting out the user message

schultz-it-solutions committed:
Update remember.php

schultz-it-solutions committed:
Update en-GB.plg_system_remember.ini

zero-24 commented on plugins/system/remember/remember.xml:
```diff
@@ -16,4 +16,21 @@
+					type="radio" 
+					label="PLG_SYSTEM_REMEMBER_RESET_LABEL"
+					description="PLG_SYSTEM_REMEMBER_RESET_DESC"
+					class="btn-group" 
```
`class="btn-group btn-group-yesno"` please

schultz-it-solutions committed:
btn-group-yesno

schultz-it-solutions committed:
Update remember.php

schultz-it-solutions committed:
Update remember.php

brianteeman commented on plugins/system/remember/remember.php:
```diff
@@ -97,8 +97,7 @@ public function onUserLogout($user, $options)
-	 * If activated in the configuration of the rememeber-me plugin, this method resets all #__user_keys for the current user
-	 * when the user changes his/her password, leaving any existing remember-me cookies on any devices useless!
+	 * Invalidate all existing remember-me cookies after a password change
 	 * This functionality was sadly inspired by the horrific events around Alice Ruggles death 
```
While I fully understand and appreciate the circumstances that caused you to create this PR I am not in favour of this comment in the code. There have been other times in the past where something similar could have been done and we have not. The correct place to "note" this issue and Joomla's positive response to it would be in the release notes, wiki, documentation etc but not in the software itself.

brianteeman commented on plugins/system/remember/remember.php:
```diff
@@ -94,4 +94,65 @@ public function onUserLogout($user, $options)
+	/**
+	 * Method is called before user data is stored in the database
+	 * Invalidate all existing remember-me cookies after a password change
+	 * This functionality was sadly inspired by the horrific events around Alice Ruggles death 
```
While I fully understand and appreciate the circumstances that caused you to create this PR I am not in favour of this comment in the code. There have been other times in the past where something similar could have been done and we have not. The correct place to "note" this issue and Joomla's positive response to it would be in the release notes, wiki, documentation etc but not in the software itself.

schultz-it-solutions replied:
If you suggest we mention this fact in the "release news" of the deployment version, I could agree to this and remove the comment from the code.

schultz-it-solutions replied:
I would classify this PR as a "feature" (rather than a "bug fix", or "security fix"). do you agree?
If so, I suggest the following text in a "feature" section of the release-news
(see e.g.: https://www.joomla.org/announcements/release-news/5710-joomla-3-7-4-release.html )

[] Invalidate all existing remember-me cookies after a password change #17827 
This functionality was sadly inspired by the horrific events around Alice Ruggles' death

Would you agree to this? or should we even make clearer that Joomla! itself was NOT in the center of named events?

mbabker replied:
If anything it would be appropriate to discuss what inspired this change as a community magazine article.  To be honest, though well intended, pointing to this fact "prominently" either in the code or a release announcement is probably not something we really want to be doing.

schultz-it-solutions replied:
Although I understand your hesitation, it looks to me as if there is no acutal guideline to handle requests for references like this one (more like personal opinons). Maybe this is something the JDT should eventually address...
In order to not further delay this PR, I am (not fully agreeing to your opposition) revoking my request for this reference, even if I still strongly belief this would be appropriate in some (to be agreed on) format.

schultz-it-solutions committed:
reference to Alice Ruggles removed!

Bakual commented:
Do you guys think we need that option? Is there a reason you **do not** want to reset the cookies after a password change? Imho that "feature" should always be enabled. I don't see why someone would want to disable it. So the option could be removed to make the code simpler.

brianteeman commented:
Agreed. Why would you want to enable a bad option

ggppdk commented:
> Do you guys think we need that option? Is there a reason you do not want to reset the cookies after a password change? Imho that "feature" should always be enabled. I don't see why someone would want to disable it. So the option could be removed to make the code simpler.

I was reading this from top, and was going to write that an option should not be added, 
this behaviour should not be possible to disable !

schultz-it-solutions committed:
making it mandatory

schultz-it-solutions committed:
Update remember.php

schultz-it-solutions committed:
making it mandatory

schultz-it-solutions committed:
making it mandatory

schultz-it-solutions committed:
making it mandatory

schultz-it-solutions commented:
is there anything left to be done before this PR gets merged back?

ghost commented:
@schultz-it-solutions at least it needs 2 successfully Tests.

schultz-it-solutions commented:
@Quy , @dgt41 , @ggppdk 
any chance one of you guys can perform another test in order to finalize this pr?

Quy commented on plugins/system/remember/remember.php:
```diff
@@ -94,4 +94,56 @@ public function onUserLogout($user, $options)
+		{
+			return true;
+		}
+		
```
Remove tabs.

Quy commented on plugins/system/remember/remember.php:
```diff
@@ -94,4 +94,56 @@ public function onUserLogout($user, $options)
+		{
+			return true;
+		}
+		
```
Remove tabs.

Quy commented on plugins/system/remember/remember.php:
```diff
@@ -94,4 +94,56 @@ public function onUserLogout($user, $options)
+				'security'
+			);
+		}
+		
```
Remove tabs.

Quy commented on plugins/system/remember/remember.php:
```diff
@@ -94,4 +94,56 @@ public function onUserLogout($user, $options)
+		
+		return true;
+	}
+	
```
Remove tab.

Quy commented on plugins/system/remember/remember.php:
```diff
@@ -94,4 +94,56 @@ public function onUserLogout($user, $options)
+		/*
+		 * But now, we need to do something 
+		 * Delete all tokens for this user!
+		*/
```
Align `*` with the other `*`.

Quy commented on plugins/system/remember/remember.php:
```diff
@@ -94,4 +94,56 @@ public function onUserLogout($user, $options)
+		try
+		{
+			$db->setQuery($query)->execute();
+		}		
```
Remove tabs after `}`.

schultz-it-solutions committed:
as per the remarks of Quy

changed

Quy commented on plugins/system/remember/remember.php:
```diff
@@ -94,4 +94,56 @@ public function onUserLogout($user, $options)
+
+		return true;
+	}
+
```
Remove blank line.

Quy commented on plugins/system/remember/remember.php:
```diff
@@ -94,4 +94,56 @@ public function onUserLogout($user, $options)
+		{
+			// Log an alert for the site admin
+			JLog::add(
+				sprintf('Failed to delete cookie token for user %s with the following error: %s', $results[0]->user_id, $e->getMessage()),
```
There is no `$results`. Use `$user['username']`.

schultz-it-solutions committed:
changed as per Quy's remarks

Quy commented:
I have tested this item :white_check_mark: successfully on f226416229b2e3bd01518529265f6ae4cc96b709<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17827">issues.joomla.org/tracker/joomla-cms/17827</a>.</sub>

schultz-it-solutions commented:
@dgt41 , @ggppdk
any chance one of you guys can perform another test in order to finalize this pr?

csthomas commented:
I have tested this item :white_check_mark: successfully on f226416229b2e3bd01518529265f6ae4cc96b709<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17827">issues.joomla.org/tracker/joomla-cms/17827</a>.</sub>

ghost commented:
Ready to Commit after two successful tests.

zero-24 commented on plugins/system/remember/remember.php:
```diff
@@ -94,4 +94,55 @@ public function onUserLogout($user, $options)
+		$db = JFactory::getDbo();
+		$query = $db->getQuery(true)
+			->delete('#__user_keys')
+			->where($db->quoteName('user_id') . ' = ' . $db->quote($user['username']));
```
hmm how can this work? Isn't the user_id just the ID and not the username?

Quy replied:
`user_id` is the `username`. The column name is misleading.

Quy replied:
For reference, I logged in as `admin` and the `user_id` is `admin` in the database.

zero-24 replied:
Ah got confused by that naming.

zero-24 commented:
Merged thanks!
